### PR TITLE
  Always use resolved file to figure out subModule name in package id

### LINKED ISSUE #30429: project references out of memory crash
**TypeScript Version:**  3.4.0-dev.20190315

**Search Terms:** project references composite oom

**Code**

https://github.com/leemhenson/project-refs-oom-repro

**Expected behavior:**

After doing the yarn install dance, `cd stacks/project-one && tsc -b --verbose` should complete without any issues.

**Actual behavior:**

`cd stacks/project-one && tsc -b --verbose` hangs and gradually chomps up enough memory to stack overflow. The cause is the `Option<Date>` annotation here:

https://github.com/leemhenson/project-refs-oom-repro/blob/master/stacks/project-one/src/app.ts#L5

**Related Issues:**

https://github.com/Microsoft/TypeScript/issues/25023
https://github.com/Microsoft/TypeScript/issues/29597

### PR DESCRIPTION
**Pull request by sheetalkamat**
Previously we calculated PackageId ( that is module name and submodule name) before resolving module and hence we didn't know when resolving `lib/option` whether it is going to be `lib\option.d.ts` file or `lib\option\index.d.ts` which resulted in us assuming first path when resolved using relative path and when resolved using non relative path respectively. So we could refer to same file but derive different package id, resulting in us assuming those as distinct path (esp if one happens to resolve in project directory and other in shared directory).
Now the submodule name is always calculated using resolved file name so as to be consistent irrespective of module name being relative or non relative.
Fixes #30429

### CODE REVIEW

File: src/compiler/moduleNameResolver.ts [hunk 1]
```diff
@@ -16,12 +16,23 @@ namespace ts {
         push(value: T): void;
     }
 
-    function withPackageId(packageId: PackageId | undefined, r: PathAndExtension | undefined): Resolved | undefined {
+    function withPackageId(packageInfo: PackageJsonInfo | undefined, r: PathAndExtension | undefined): Resolved | undefined {
+        let packageId: PackageId | undefined;
+        if (r && packageInfo) {
+            const packageJsonContent = packageInfo.packageJsonContent as PackageJson;
+            if (typeof packageJsonContent.name === "string" && typeof packageJsonContent.version === "string") {
```
Comment (sandersn) [inline src/compiler/moduleNameResolver.ts]: `name` and `version` have type `string | undefined`, right? It seems a lot clearer to write `packageJsonContent.name !== undefined`

Comment (sandersn) [reply src/compiler/moduleNameResolver.ts]: If name or version aren't defined, then you get no packageId now. Before this change, was it possible to get a packageId in situations where name or version would be undefined? What happens now in that case (if it exists)?

File: tests/baselines/reference/moduleResolutionWithExtensions_unexpected.trace.json [hunk 2]
```diff
@@ -2,11 +2,8 @@
     "======== Resolving module 'normalize.css' from '/a.ts'. ========",
     "Module resolution kind is not specified, using 'NodeJs'.",
     "Loading module 'normalize.css' from 'node_modules' folder, target file type 'TypeScript'.",
-    "'package.json' does not have a 'typings' field.",
-    "'package.json' does not have a 'types' field.",
```
Comment (sandersn) [inline tests/baselines/reference/moduleResolutionWithExtensions_unexpected.trace.json]: why do these lines go away?

File: src/compiler/moduleNameResolver.ts (hunk 1 above)
Comment (sheetalkamat) [reply src/compiler/moduleNameResolver.ts]: it was exact same condition for package id to be present .. just copied it here instead. Given that the value is coming from Json I think checking string type is better but check can be updated to isString() instead for better readability

File: tests/baselines/reference/moduleResolutionWithExtensions_unexpected.trace.json (hunk 2 above)
Comment (sheetalkamat) [reply tests/baselines/reference/moduleResolutionWithExtensions_unexpected.trace.json]: Because we don’t look at the field to calculate sub module names

Comment (sheetalkamat) [reply tests/baselines/reference/moduleResolutionWithExtensions_unexpected.trace.json]: Previously when sub module name was empty string these fields were read to get better name

File: src/compiler/moduleNameResolver.ts (hunk 1 above)
Comment (sandersn) [reply src/compiler/moduleNameResolver.ts]: No, I think `typeof x==='string'` is just as readable as `isString`.

**Pull request merged**